### PR TITLE
conntrack module

### PR DESCRIPTION
**Pull request by mcluseau**

### CODE REVIEW

File: client/plugins/conntrack/conntrack_linux.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	return parameters
+}
+
+func protoStr(proto v1.Protocol) string {
+	return strings.ToLower(proto.String())
+}
+
+// IsClearConntrackNeeded returns true if protocol requires conntrack cleanup for the stale connections
+func IsClearConntrackNeeded(proto v1.Protocol) bool {
+	return proto == v1.Protocol_UDP || proto == v1.Protocol_SCTP
```
Comment (knabben) [inline client/plugins/conntrack/conntrack_linux.go:67]: why do need to cleanup SCTP conntrack?

File: client/plugins/conntrack/conntrack_linux.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	"k8s.io/utils/exec"
+	utilnet "k8s.io/utils/net"
+
+	v1 "sigs.k8s.io/kpng/api/localnetv1"
+)
+
+var execer = exec.New()
+
+func setupConntrack() {
+	// TODO
```
Comment (knabben) [inline client/plugins/conntrack/conntrack_linux.go:17]: do we need to setup anything before calling the conntrack module, or this is only a placeholder?

File: client/plugins/conntrack/conntrack_linux.go (hunk 1 above)
Comment (mcluseau) [reply client/plugins/conntrack/conntrack_linux.go:67]: this code is from upstream ;)

File: client/plugins/conntrack/conntrack_linux.go (hunk 2 above)
Comment (mcluseau) [reply client/plugins/conntrack/conntrack_linux.go:17]: there are things to do yes (upstream: cmd/kube-proxy/app/conntrack.go)

File: backends/nft/register.go [hunk 3]
```diff
@@ -25,7 +27,12 @@ func (b *backend) Sink() localsink.Sink {
 	sink := fullstate.New(&b.cfg)
 
 	PreRun()
-	sink.Callback = Callback
+
+	ct := conntrack.New()
+	sink.Callback = fullstatepipe.New(fullstatepipe.ParallelSendSequenceClose,
+		Callback,
+		ct.Callback,
```
Comment (jayunit100) [inline backends/nft/register.go:34]: this is a new "mini backend" that is executed after the original , NFT backend

File: client/plugins/conntrack/conntrack.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+					}
+				}
+			}
+		}
+	}
+
+	ct.flows.Done()
+
+	for _, item := range ct.flows.Deleted() {
+		cleanupFlowEntries(item.Value().Get())
```
Comment (VivekThrivikraman-est) [inline client/plugins/conntrack/conntrack.go:58]: I think cleaning flow entries should also be done when endpoint changes from no endpoint to >=1  endpoints as there could be stale entries then.
We should also be checking the ready state of endpoints , else there are chances of stale connections(even if we take care of transition from 0 to 1 endpoints)

File: backends/nft/register.go [hunk 5]
```diff
@@ -25,7 +27,12 @@ func (b *backend) Sink() localsink.Sink {
 	sink := fullstate.New(&b.cfg)
 
 	PreRun()
-	sink.Callback = Callback
+
+	ct := conntrack.New()
+	sink.Callback = fullstatepipe.New(fullstatepipe.ParallelSendSequenceClose,
```
Comment (VivekThrivikraman-est) [inline backends/nft/register.go:32]: I think we should be calling the tasks in serial order 1)Apply the rules 2)Delete the stale endpoints.

File: client/plugins/conntrack/conntrack_linux.go (hunk 1 above)
Comment (VivekThrivikraman-est) [inline client/plugins/conntrack/conntrack_linux.go:67]: I think SCTP can be actually removed

**Pull request merged**